<!-- Begin page content -->
<main role="main" class="flex-shrink-0">
    <div class="container">
        
           <h2 class="text-left" style="padding: 5px;">MutViz</h2>

        <div style="padding-left: 15px; margin-bottom: 10px;">
            <p>MutViz helps to identify <b>enrichments of somatic mutations</b> (single nucelotide variants) falling on predefined or user-specified sets of genomic regions, such as transcription factor binding sites, promoters, specific gene sets and others. <br>Users can choose among <b>several  mutation  datasets</b>  collected  from  public  repositories and  upload  <b>custom  sets  of  genomic  regions</b>  in  a  BED  file format.  <br>Moreover, MutViz provides visualization features and simple statistics that ease the comparative analysis between different types of mutations (e.g., ”C→T” is one such type) and different sets of regions and mutations. </p>
        
            
            We currently support mutations for the following tumor types:<br>

                    <select style="margin-bottom: 5px;">
                        <option ng-repeat="t in tumorTypes.available">{{t.name}}</option>
            </select>
            <br>
            The database was build starting from <a href="https://icgc.org/">ICGC</a> data (download date July 2019).
            <br><br>
            
            Please cite the following articles if you use MutViz in your research:
            <ul>

                <li>

                    <p>
                        <i>Gulino, Andrea, Eirini Stamoulakatou, and Rosario M. Piro. <br>
                        <b>MutViz 2.0: visual analysis of somatic mutations and the impact of mutational signatures on selected genomic regions.</b> <br>
                        NAR Cancer 3.2 (2021): zcab012.</i>
                    </p>
                </li>
                <li>

                    <p><i>
                        Gulino, Andrea, Eirini Stamoulakatou, Arif Canakoglu and Pietro Pinoli <br>
                        <b>Analysis and visualization of mutation enrichments for selected genomic regions and cancer types.</b><br> 2019 IEEE International Conference on Bioinformatics and Biomedicine (BIBM). IEEE, 2019.</i>
                    </p>

                 </li>
        </ul>
            <br>
            
                    <img src="img/intro-min.png" width="100%">
            
            <br><br>
            
         

            <h4>Data and Workspace</h4>
            <p>
                Before using the tool, the user must add at least one dataset to the <a href="#!/data">workspace</a>:</p>
            <ul>
                <li>Either uploading a custom file (BED or narrowpeak). E.g. of BED file: 
                    <pre style="background-color: aliceblue; padding: 15px;">chr10	102639623	102639642	*
chr10	102639678	102639697	*</pre>

                </li>
                <li>Or choosing a dataset from our public repository.<br>
                
                    <select style="margin-bottom: 5px;" ng-model="repoEl">
                        <option ng-repeat="t in repository" ng-value="t">{{t.name}}</option>
                    </select>
                    <div ng-bind-html="repoEl.description" style="background-color: #f6f6ff;"></div>

                </li>
            </ul>
            <br>
            All regions should use <b>reference genome hg19</b>. 
            <br>
            
            The workspace structure is stored on the browser's local storage and will be available until the local storage is deleted.  Therefore, closing the browser or reloading the page won't erase data in your work space. However, uploaded region sets will disappear from the workspace as soon as their survival time on our database expires (few days).


            <!--h4>Horizontal axis and interaction</h4>

<p>In all the plots the horizontal axis represents the distance of a mutation w.r.t. the center of a region. The domain can be enlarged or shrinked by using the availble range-selector component.</p>
<br>
<img src="img/slider.png" width="500">

<br><br>

<h4>Binning and interaction</h4>
Data binning is used in all the plots. The <b>bin size</b>, in base pairs, can be changed by using the form available on the left side of each plot. Bins are centered on the motif (i.e. on 0).
<br><br>
<img src="img/bin_size.png" width="200">

<br><br>
<h4>Mutation Set selector</h4>
<p>The tooldbar on the left of every plot allows to restrict the set of mutations to consider. Mutations in the defined set must be independed (e.g. you cannot choose "A > *" and "A > C"). If the set contains conditions that are not independed an error message will popup and the plot will not be updated until the defined set is fixed. </p>
<img src="img/mutations.png" width="200"-->

            <div class="text-center">
                <h2>MUTVIZ 1.0</h2>
                <hr>
            </div>

            Visualizations in this section were presented at the IEEE BIBM 2019 Conference, San Diego. You can find the related publication on <a target="_blank" href="https://ieeexplore.ieee.org/abstract/document/8983196/">IEEE Xplore</a>. <br><br>

            <h2> <a href="#!/uc1" target="_blank" style="color: #24587e;">Histogram</a></h2>
            <div class="row">
                <div class="col-md-8">
                    The visualization consists of an histogram showing the distribution of mutations around the regions in the provided region-set. The <b>x-axis</b> represents the distance from the center of the provided regions, computed as <code>(start+stop)/2</code>, and the <b>y-axis</b> represents the number of mutations falling in the corresponding bin.
                    Use the toolbar on the left to select:
                    <ul>
                        <li>A <b>region set</b>, among the ones added to the <a href="#!/data">workspace</a>.</li>
                        <li>A <b>tumor type</b>: only mutations for the selected tumor type will be displayed.</li>
                        <li><b>Mutation classes</b>: while they gray rectangles show all the mutations, colored bars represent only the selected mutation classes. </li>
                        <li>By enabling the <b>stacked</b> option the tool will represent each selected mutation class with a different color. </li>

                        <li>Disabling <b>show total</b>, the gray bars representing all the mutations will be removed from the plot.</li>
                        <li><b>bin size</b>: number of base pairs per bin.</li>
                        <li><b>min-y</b>: the mininum height for the y-axis.</li>
                    </ul>

                    The horizontal slider helps restricting the domain.<br>

                    The <b>statistical test</b> box allows to select the area to test for increased (decreased) mutation rate by specifying  the initial and final x-coordinate of the area. The testing area is displayed in the plot as a light-blue rectangle. The <b>bin-coordinate</b> column shows the bins corresponding to the selected coordinates. The p-value is the result of the performed permutation test. 
                    <b style="color:red">Only the selected mutation classes will be considered for testing.</b><br>
                    For further details, please refer to our  <a target="_blank" href="https://ieeexplore.ieee.org/abstract/document/8983196/"> 2019 IEEE BIBM paper</a>.

                </div>
                <div class="col-md-4 text-center" >
                    <div style="background-image: url('img/example1.png');background-repeat: no-repeat; background-position: center; background-size: contain;">

                        <div style="background-color: #dedede1c;  height: 200px; line-height: 200px;">
                            <!--a href="index.html#!/uc1?showExample=1" class="btn btn-primary" style=" vertical-align: middle;">Load Example</a-->
                        </div>
                    </div>
                </div>


            </div>

            <br><br>

            <h2><a href="#!/uc2" target="_blank" style="color: #24587e;">Region comparison</a></h2>
            <div class="row">
                <div class="col-md-8">

                    The visualization allows to compare the distribution of mutations around regions in <b>two different region-sets</b> through a heat-map. The <b>x-axis</b> represents the distance from the center of the provided regions, computed as <code>(start+stop)/2</code>, the <b>y-axis</b> discriminates the two selected region sets and the color-intensity represents the amount of mutations falling in a given bin. The most opaque bin for a region set represents the bin containing the highest number of mutations.
                    Use the toolbar on the left to select:
                    <ul>
                        <li><b>Two region sets</b>, among the ones added to the <a href="#!/data">workspace</a>.</li>
                        <li>A <b>tumor type</b>: only mutations for the selected tumor type will be displayed.</li>
                        <li><b>Mutation classes</b>: only mutations belonging to the selected classes will be displayed. </li>
                        <li><b>bin size</b>: number of base pairs per bin.</li>
                    </ul>

                    The horizontal slider helps restricting the domain.<br/>

                    The <b>statistical test</b> returns a p-value indicating how different the currently displayed distributions are. <b style="color:red">Only the selected mutation classes will be considered for testing.</b><br>
                    For further details, please refer to our  <a target="_blank" href="https://ieeexplore.ieee.org/abstract/document/8983196/"> 2019 IEEE BIBM paper</a>.

                </div>
                <div class="col-md-4 text-center" >
                    <div style="background-image: url('img/example2.png');background-repeat: no-repeat; background-position: center; background-size: contain;">

                        <div style="background-color: #dedede1c;  height: 200px; line-height: 200px;">
                            <!--a href="index.html#!/uc2?showExample=1" class="btn btn-primary" style=" vertical-align: middle;">Load Example</a-->
                        </div>
                    </div>
                </div>
            </div>

            <br><br>

            <h2><a href="#!/uc3" target="_blank" style="color: #24587e;">Tumor comparison</a></h2>
            <div class="row">
                <div class="col-md-8">
                    The visualization allows to compare the distribution of <b>mutations from different tumor-types</b> around regions in two provided region-set through a heat-map. The <b>x-axis</b> represents the distance from the center of the provided regions, computed as <code>(start+stop)/2</code>, the <b>y-axis</b> discriminates the selected tumor-types and the color-intensity represents the amount of mutations falling in a given bin. The most opaque bin for a region set represents the bin containing the highest number of mutations.
                    Use the toolbar on the left to select:
                    <ul>
                        <li>A <b>region set</b>, among the ones added to the <a href="#!/data">workspace</a>.</li>
                        <li><b>Multiple tumor types</b>: each row of the heatmap shows the distribution of mutations for a given tumor type.</li>
                        <li><b>Mutation classes</b>: only mutations belonging to the selected classes will be displayed. </li>
                        <li><b>bin size</b>: number of base pairs per bin.</li>
                    </ul>

                    The horizontal slider helps restricting the domain.<br/>

                    The <b>statistical test</b> returns a p-value indicating how different the currently displayed distributions are.. <b style="color:red">Only the selected mutation classes will be considered for testing.</b><br>
                    For further details, please refer to our  <a target="_blank" href="https://ieeexplore.ieee.org/abstract/document/8983196/"> 2019 IEEE BIBM paper</a>.
                </div>
                <div class="col-md-4 text-center" >
                    <div style="background-image: url('img/example3.png');background-repeat: no-repeat; background-position: center; background-size: contain;">

                        <div style="background-color: #dedede1c;  height: 200px; line-height: 200px;">
                            <!--a href="index.html#!/uc3?showExample=1" class="btn btn-primary" style=" vertical-align: middle;">Load Example</a-->
                        </div>
                    </div>
                </div>
            </div>
        </div>


        <div class="text-center">
            <h2>MUTVIZ 2.0</h2>
            <hr>
        </div>
        
        Differently from the visualiations in MutViz 1.0 (histogram, regions and tumors), this representation takes into account only mutations overlapping with regions in the provided region set: i.e., given the set of mutations M and the region set R, the considered set of mutations is:
             
                    <div class="text-center">
                        <code>{m in M : &exist;r in R, m.position &ge; r.start AND m.position &le; r.stop} </code>
                    </div>


        <br><br>
        
        <h2><a href="#!/uc4" target="_blank" style="color: #24587e;">Trinucleotide mutations</a></h2>
            <div class="row">
                <div class="col-md-8">
                   The visualization consists of a bar-plot representing the distribution of the six classes of base substitutions (C→A,C→G, C→T, T→A, T→C, T→G) within their immediate sequence context which fall into the provided region set. <br><br>
                    
                    This visualization takes into account the  information  from  the  5’  and  3’  adjacent  bases  (also called  flanking  base  pairs  or  trinucleotide  context)  leading  to <b>96</b> possible mutation types (e.g. A[C→A]A, A[C→A]T, etc.). The  mutation  catalog  of  a  tumor  is  created  by  categorizing each  single  nucleotide  variant  (SNV)  (synonyms:  base-pair-substitution  or  substitution  point  mutation)  in  one  of  the  96 mutation types and counting the total number of substitutions for each of these 96 mutation types.<br><br>

                    The <b>horizontal-axis</b> discriminates the  96 mutation types and the <b>y-axis</b> represents the number of mutations for that type falling onto the region-set. Plots are grouped by the class of base substitution.<br>
                    Use the toolbar on the left to select: <br>
                    <ul>
                        <li>A <b>region set</b>, among the ones added to the <a href="#!/data">workspace</a>.</li>
                        <li>A <b>tumor type</b>: only mutations for the selected tumor type will be displayed. You can further <b>filter mutations</b> by providing a predicate on the <b>clinical condition</b> of donors associated to the selected tumor type. Click on the filter-button next to the tumor-type selector to open the predicate editor. On top of the editor you will see the number of donors matching the current predicate.</li>
                        <li><b>Mutation classes</b>: only mutations belonging to the selected classes will be displayed. </li>
                    </ul>
                    
                </div>
                <div class="col-md-4 text-center" >
                    <div style="background-image: url('img/example4.png');background-repeat: no-repeat; background-position: center; background-size: contain;">

                        <div style="background-color: #dedede1c;  height: 200px; line-height: 200px;">
                            <!--a href="index.html#!/uc1?showExample=1" class="btn btn-primary" style=" vertical-align: middle;">Load Example</a-->
                        </div>
                    </div>
                </div>


            </div>

            <br><br>
        
        <h2><a href="#!/uc5" target="_blank" style="color: #24587e;">Mutations per donor</a></h2>
            <div class="row">
                <div class="col-md-8">
                    The visualization shows a box plot for each class of base substitution (C→A,C→G, C→T, T→A, T→C, T→G), representing the distribution of <b>mutations per-donor</b> overlapping regions in the provided region set. The box-plots on the right group the selected mutation types into transitions (Ti) and transversions (Tv).<br>

                    When the <b>show trinucleotide</b> option is enabled, the visualization takes into account the  information  from  the  5’  and  3’  adjacent  bases  (also called  flanking  base  pairs  or  trinucleotide  context)  leading  to <b>96</b> possible mutation types (e.g. A[C→A]A, A[C→A]T, etc.). The  mutation  catalog  of  a  tumor  is  created  by  categorizing each  single  nucleotide  variant  (SNV)  (synonyms:  base-pair-substitution  or  substitution  point  mutation)  in  one  of  the  96 mutation types and counting the total number of substitutions for each of these 96 mutation types.

                    The <b>horizontal-axis</b> discriminates the 6 (96) mutation types (trinucleotide types) and the  <b>vertical-axis</b> represents the number of mutations per-donor, of that type, overlapping regions in the region-set.<br>
                    Hovering the box-plot with your mouse you will be able to get more information on the distribution and on outliers (e.g. you can get the donor-id associated to an outlier).
                    Use the toolbar on the left to select:<br>
                    <ul>
                        <li>A <b>region set</b>, among the ones added to the <a href="#!/data">workspace</a>.</li>
                        <li>A <b>tumor type</b>: only mutations for the selected tumor type will be displayed. You can further <b>filter mutations</b> by providing a predicate on the <b>clinical condition</b> of donors associated to the selected tumor type. Click on the filter-button next to the tumor-type selector to open the predicate editor. On top of the editor you will see the number of donors matching the current predicate.</li>
                        <li><b>Mutation classes</b>: only mutations belonging to the selected classes will be displayed. </li>
                    </ul>


                    
                </div>
                <div class="col-md-4 text-center" >
                    <div style="background-image: url('img/example5.png');background-repeat: no-repeat; background-position: center; background-size: contain;">

                        <div style="background-color: #dedede1c;  height: 200px; line-height: 200px;">
                            <!--a href="index.html#!/uc1?showExample=1" class="btn btn-primary" style=" vertical-align: middle;">Load Example</a-->
                        </div>
                    </div>
                </div>


            </div>

            <br><br>
        
        <h2><a href="#!/uc6" target="_blank" style="color: #24587e;">Signature Refitting</a></h2>
            <div class="row">
                <div class="col-md-8">
                    Given a catalogue of mutational <b>signatures</b> S and a set of somatic mutations M found in a given tumor genome, exposures E can be computed ("signature refitting"). The exposures predict what fractions of the mutations can be attributed to the given signatures. Exposures are computed such that the exposure-weighted sum of signatures reflects the distribution of mutation types in the tumor genome.<br><br>

                    Here, the set of somatic mutations M used are only those mutations which fall onto the region set specified by the user. Since, however, the region set may have a largely different trinucleotide content than the whole genome, for which mutational signatures have been originally been determined, the signatures are adjusted according to the trinucleotide content of the region set prior to determining exposures.<br><br>
                    
                    This visualization shows a <b>bar-plot</b> in which, for each signature (horizontal-axis), the corresponding exposure (verical-axis) is reported. Alternatively, a <b>box-plot</b> for each signature shows the distribution of exposures over donors.<br><br>
                    
                    For higher accuracy in signature refitting, when the <b>threshold</b> is enabled (e.g. <code>threshold=N</code>), the tool will consider only tumors (synonyms: donors, patients) for wich more than N mutations overlap regions in the provided region-set. In case less than 5 tumors have more than N mutations, the mutations are merged and an estimation of the average signature exposure is provided.<br><br>
                           
                    In case the box-plot visualization is selected, hovering the box-plot with your mouse you will be able to get more information on the distribution and on outliers (e.g. you can get the donor-id associated to an outlier).<br><br>
                    Use the toolbar on the left to select:<br>
                    <ul>
                        <li>A <b>region set</b>, among the ones added to the <a href="#!/data">workspace</a>.</li>
                        <li>A <b>tumor type</b>: only mutations for the selected tumor type will be displayed. You can further <b>filter mutations</b> by providing a predicate on the <b>clinical condition</b> of donors associated to the selected tumor type. Click on the filter-button next to the tumor-type selector to open the predicate editor. On top of the editor you will see the number of donors matching the current predicate.</li>
                        <li>A <b>threshold</b> (N): when enabled and set, only donors (tumors) with more than N mutations overlapping the region-set will be considered.</li>
                        <li><b>Mutation classes</b>: only mutations belonging to the selected classes will be displayed. </li>
                    </ul>
                    <br><br>
                    
                    Once the plot is ready the number of considered patients will be displayed in the toolbar. <br><br>
                    
                    Source of mutational signatures: <a target="_blank" href="https://cancer.sanger.ac.uk/cosmic/signatures">COSMIC Mutational Signatures v3</a> and <a  target="_blank"  href="https://pubmed.ncbi.nlm.nih.gov/32025018/">Alexandrov et al</a>.<br><br>
                    
                    We considered only single base substitution (SBS) signatures, excluding those which have been characterized as possible sequencing artifacts (SBS 27, 29, 43,  and 45--60) or identified only in exome sequencing samples (SBS 23, 25,  and 42).
                </div>
                <div class="col-md-4 text-center" >
                    <div style="background-image: url('img/example6.png');background-repeat: no-repeat; background-position: center; background-size: contain;">

                        <div style="background-color: #dedede1c;  height: 200px; line-height: 200px;">
                            <!--a href="index.html#!/uc1?showExample=1" class="btn btn-primary" style=" vertical-align: middle;">Load Example</a-->
                        </div>
                    </div>
                </div>


            </div>

            <br><br>

    </div>
</main>
